File: xuri-rpc-java/src/main/java/com/xuri/rpc/Rpc.java
package com.xuri.rpc;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicLong;

/**
 * RPC框架核心工具类，包含全局状态和静态工具方法
 * 对应TS中 rpc.ts 模块级别的变量和函数
 */
public final class Rpc {

    private Rpc() {
    }

    // ==================== 全局状态 ====================

    /** 调试标志, 对应TS: let debugFlag = false */
    private static volatile boolean debugFlag = false;

    /** 本地主机ID, 对应TS: let hostId: string | null = null */
    private static volatile String hostId = null;

    /** 全局消息接收器单例, 对应TS: let messageReceiver: MessageReceiver | null */
    private static volatile MessageReceiver messageReceiver = null;

    /** ID计数器 (线程安全), 对应TS: let idCount = 0 */
    private static final AtomicLong idCount = new AtomicLong(0);

    /** 增强类型标志, 对应TS: let enchanceType = true */
    private static boolean enhanceType = true;

    /** GC超时限制(毫秒), 对应TS: const gcTimeLimit = 30 * 1000 */
    static final long GC_TIME_LIMIT = 30 * 1000;

    /** 连接超时限制(毫秒), 对应TS: const connectionLimit = 30 * 1000 */
    static final long CONNECTION_LIMIT = 30 * 1000;

    /**
     * 每个hostId的选项映射, 对应TS: let options: Record<string|symbol, MessageReceiverOptions> = {}
     * 注意: TS中使用Symbol('defaultHost')作为默认key，Java中使用null key由DEFAULT_HOST_KEY替代
     */
    private static final ConcurrentHashMap<String, MessageReceiverOptions> options = new ConcurrentHashMap<>();

    /** 默认host的key, 对应TS: let defaultHost = Symbol('defaultHost') */
    private static final String DEFAULT_HOST_KEY = "__DEFAULT_HOST__";

    // ==================== 浅自动包装器 ====================

    /**
     * 浅层自动包装器, 对应TS: const shallowAutoWrapper: AutoWrapper
     *
     * 规则:
     * - null → 原样返回
     * - RpcFunction → 包装为代理
     * - List中包含RpcFunction → 整个List包装为代理
     * - Map中包含RpcFunction值 → 整个Map包装为代理
     * - 其他 → 原样返回
     *
     * 注意: TS中检查 typeof obj == 'function'，Java中对应为 instanceof RpcFunction
     * (因为Java没有一等函数类型，需要通过接口标识可调用对象)
     */
    public static final AutoWrapper SHALLOW_AUTO_WRAPPER = (Object obj) -> {
        if (obj == null) {
            return null;
        }
        // 对应TS: typeof obj == 'function'
        if (obj instanceof MessageReceiver.RpcFunction) {
            return asProxy(obj);
        }
        if (obj instanceof List) {
            List<?> list = (List<?>) obj;
            boolean notPureData = false;
            for (Object item : list) {
                if (item instanceof MessageReceiver.RpcFunction) {
                    notPureData = true;
                    break;
                }
            }
            return notPureData ? asProxy(obj) : obj;
        }
        if (obj instanceof Map) {
            @SuppressWarnings("unchecked")
            Map<String, Object> map = (Map<String, Object>) obj;
            boolean notPureData = false;
            for (Object value : map.values()) {
                if (value instanceof MessageReceiver.RpcFunction) {
                    notPureData = true;
                    break;
                }
            }
            return notPureData ? asProxy(obj) : obj;
        }
        return obj;
    };

    // ==================== 公共方法 ====================

    /**
     * 设置调试标志
     * 对应TS: export function setDebugFlag(flag: boolean)
     */
    public static void setDebugFlag(boolean flag) {
        debugFlag = flag;
    }

    public static boolean isDebug() {
        return debugFlag;
    }

    /**
     * 设置主机ID
     * 对应TS: export function setHostId(id: string)
     */
    public static void setHostId(String id) {
        hostId = id;
        getOrCreateOption(null).hostId = id;
    }

    public static String getHostId() {
        return hostId;
    }

    /**
     * 按ID删除代理
     * 对应TS: export function _deleteProxyById(id: string, hostId?: string)
     */
    public static void deleteProxyById(String id, String hostIdParam) {
        getOrCreateOption(hostIdParam).objectOfProxyManager.deleteById(id);
    }

    /**
     * 按对象删除代理
     * 对应TS: export function _deleteProxy(obj: object, hostId?: string)
     */
    public static void deleteProxy(Object obj, String hostIdParam) {
        getOrCreateOption(hostIdParam).objectOfProxyManager.delete(obj);
    }

    /**
     * 生成唯一ID
     * 对应TS: function getId() { return hostId + '' + (idCount++) }
     */
    public static String getId() {
        return (hostId != null ? hostId : "") + idCount.getAndIncrement();
    }

    /**
     * 获取或创建全局消息接收器
     * 对应TS: export function getMessageReceiver(): MessageReceiver
     */
    public static MessageReceiver getMessageReceiver() {
        if (messageReceiver == null) {
            synchronized (Rpc.class) {
                if (messageReceiver == null) {
                    messageReceiver = new MessageReceiver();
                }
            }
        }
        return messageReceiver;
    }

    /**
     * 获取或创建指定hostId的选项
     * 对应TS: function getOrCreateOption(id?: string | null | symbol): MessageReceiverOptions
     */
    static MessageReceiverOptions getOrCreateOption(String id) {
        if (id == null) {
            id = DEFAULT_HOST_KEY;
        }
        if (id.equals(hostId)) {
            id = DEFAULT_HOST_KEY;
        }
        final String key = id;
        return options.computeIfAbsent(key, k -> {
            String resolvedHostId = DEFAULT_HOST_KEY.equals(k) ? null : k;
            return new MessageReceiverOptions(resolvedHostId);
        });
    }

    // ==================== 代理相关 ====================

    /**
     * 获取或生成对象的代理ID
     * 对应TS: function getOrGenerateObjectId(obj: object, hostIdFrom: string)
     */
    private static String getOrGenerateObjectId(Object obj, String hostIdFrom) {
        ObjectOfProxyManager proxyManager = getOrCreateOption(hostIdFrom).objectOfProxyManager;
        if (hostId == null) {
            throw new RuntimeException("hostId is null");
        }
        if (!proxyManager.has(obj)) {
            String id = getId();
            proxyManager.set(obj, id);
        }
        return proxyManager.get(obj);
    }

    /**
     * 为对象创建代理描述
     * 对应TS: function createProxyForObject(proxyId, obj, hostId)
     *
     * 在TS中检查typeof obj == 'function'和Object.keys过滤函数属性;
     * Java中: RpcFunction对应函数，Map中的RpcFunction值对应函数属性
     */
    private static ProxyDescriber createProxyForObject(String proxyId, Object obj, String hostIdParam) {
        if (obj == null) {
            return null;
        }

        ProxyDescriber proxy = new ProxyDescriber();
        proxy.setId(proxyId);
        proxy.setHostId(hostIdParam);

        List<ProxyDescriber.MemberDescriber> members = new ArrayList<>();

        // 对应TS: if(typeof obj == 'function')
        if (obj instanceof MessageReceiver.RpcFunction) {
            members.add(new ProxyDescriber.MemberDescriber("function", "__call__"));
        } else if (obj instanceof Map) {
            // 对应TS: Object.keys(obj).filter(k => typeof obj[k] == 'function').filter(k => !k.startsWith('__'))
            @SuppressWarnings("unchecked")
            Map<String, Object> map = (Map<String, Object>) obj;
            for (Map.Entry<String, Object> entry : map.entrySet()) {
                if (entry.getValue() instanceof MessageReceiver.RpcFunction
                        && !entry.getKey().startsWith("__")) {
                    members.add(new ProxyDescriber.MemberDescriber("function", entry.getKey()));
                }
            }
        }
        // 注意: 对于非Map非RpcFunction的Java对象，此处不会生成members
        // 在实际Java使用中，可通过反射获取public方法列表 (留作扩展)

        proxy.setMembers(members);
        return proxy;
    }

    /**
     * 将对象包装为代理PreArgObj
     * 对应TS: export function asProxy(obj: object, hostIdFrom?: string): PreArgObj
     */
    public static PreArgObj asProxy(Object obj, String hostIdFrom) {
        String resolvedHostId = getOrCreateOption(hostIdFrom).hostId;
        String id = getOrGenerateObjectId(obj, resolvedHostId);
        ProxyDescriber proxy = createProxyForObject(id, obj, resolvedHostId);
        return new PreArgObj("proxy", proxy);
    }

    public static PreArgObj asProxy(Object obj) {
        return asProxy(obj, null);
    }

    /**
     * 生成错误响应
     * 对应TS: export function generateErrorReply(message, errorText, status = 500)
     */
    public static Response generateErrorReply(Request message, String errorText, int status) {
        Response reply = new Response();
        reply.setId(getId());
        reply.setIdFor(message.getId());
        reply.setTrace(errorText);
        reply.setStatus(status);
        return reply;
    }

    public static Response generateErrorReply(Request message, String errorText) {
        return generateErrorReply(message, errorText, 500);
    }

    // ==================== 定时任务 ====================

    /**
     * 启动自动检查定时任务
     * 对应TS: export function autoCheck()
     *
     * 每3秒执行:
     * 1. killTimeoutConnection - 清理超时连接
     * 2. autoReRegister - 自动重新注册
     * 3. removeOutdatedProxyObject - 移除过期代理
     */
    public static void autoCheck() {
        ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor(r -> {
            Thread t = new Thread(r, "rpc-auto-check");
            t.setDaemon(true);
            return t;
        });

        scheduler.scheduleAtFixedRate(Rpc::killTimeoutConnection, 3, 3, TimeUnit.SECONDS);
        scheduler.scheduleAtFixedRate(Rpc::autoReRegister, 3, 3, TimeUnit.SECONDS);
        scheduler.scheduleAtFixedRate(() -> removeOutdatedProxyObject(-1), 3, 3, TimeUnit.SECONDS);
    }

    /**
     * 清理超时连接
     * 对应TS: function killTimeoutConnection(client: Client | null = null, millSec: number = -1)
     */
    public static void killTimeoutConnection() {
        killTimeoutConnection(null, -1);
    }

    public static void killTimeoutConnection(Client client, long millSec) {
        if (millSec == -1) {
            millSec = CONNECTION_LIMIT;
        }
        final long timeout = millSec;

        if (client != null) {
            processTimeoutForOption(getOrCreateOption(client.getHostId()), timeout);
        } else {
            for (MessageReceiverOptions option : options.values()) {
                processTimeoutForOption(option, timeout);
            }
        }
    }

    private static void processTimeoutForOption(MessageReceiverOptions option, long timeout) {
        ConcurrentHashMap<String, PendingRequest> reqPending = option.requestPendingDict;
        List<String> toRemove = new ArrayList<>();
        for (Map.Entry<String, PendingRequest> entry : reqPending.entrySet()) {
            PendingRequest pending = entry.getValue();
            if (System.currentTimeMillis() - pending.sendTime > timeout) {
                pending.future.completeExceptionally(new RuntimeException("timeout"));
                toRemove.add(entry.getKey());
            }
        }
        for (String key : toRemove) {
            reqPending.remove(key);
        }
    }

    /**
     * 自动重新注册远程代理
     * 对应TS: export function autoReRegister()
     */
    public static void autoReRegister() {
        for (MessageReceiverOptions option : options.values()) {
            RemoteProxyManager manager = option.runnableProxyManager;
            for (Map.Entry<Client, Set<String>> entry : manager.getClientMap().entrySet()) {
                Client client = entry.getKey();
                Set<String> ids = entry.getValue();
                if (ids == null) {
                    continue;
                }

                List<List<Object>> toReRegister = new ArrayList<>();
                Set<String> toRemove = new HashSet<>();

                for (String id : ids) {
                    Object obj = manager.get(id);
                    if (obj == null) {
                        toRemove.add(id);
                        continue;
                    }
                    List<Object> item = new ArrayList<>();
                    item.add(id);
                    toReRegister.add(item);
                }

                // 清理已失效的id
                for (String id : toRemove) {
                    ids.remove(id);
                }
                if (ids.isEmpty()) {
                    manager.removeFromClientMap(client);
                }

                // 调用远程reRegister方法
                if (!toReRegister.isEmpty()) {
                    /* @async 原TS中 client.getObject('main0').reRegister(toReRegister)，此处同步调用 */
                    try {
                        Object main0 = client.getObject("main0");
                        if (main0 instanceof RemoteObject) {
                            ((RemoteObject) main0).invoke("reRegister", (Object) toReRegister);
                        }
                    } catch (Exception e) {
                        System.err.println("autoReRegister failed: " + e.getMessage());
                    }
                }
            }
        }
    }

    /**
     * 移除过期代理对象
     * 对应TS: export function removeOutdatedProxyObject(timeout: number = -1)
     * (已修复bug: 原TS中错误遍历 Object.entries(manager)，应遍历 manager.reverseProxyMap)
     */
    public static void removeOutdatedProxyObject(long timeout) {
        if (timeout <= 0) {
            timeout = GC_TIME_LIMIT;
        }
        for (Map.Entry<String, MessageReceiverOptions> entry : options.entrySet()) {
            ObjectOfProxyManager manager = entry.getValue().objectOfProxyManager;
            int before = manager.size();
            int count = manager.removeOutdated(timeout);
            if (count > 0 && debugFlag) {
                System.out.println(entry.getKey() + " removed " + count
                        + " proxies, before " + before + " after " + manager.size());
            }
        }
    }

    // ==================== 类型检查工具 ====================

    /**
     * 判断是否为简单/原始类型对象
     * 对应TS: function isSimpleObject(obj)
     *
     * 简单类型: null, String, Number(Integer/Long/Double/Float等), Boolean, byte[]
     */
    public static boolean isSimpleObject(Object obj) {
        return obj == null
                || obj instanceof String
                || obj instanceof Number
                || obj instanceof Boolean
                || obj instanceof byte[];
    }

    /**
     * 判断消息是否为响应
     * 对应TS: function isResponse(message): message is Response
     */
    public static boolean isResponse(Message message) {
        return message instanceof Response && ((Response) message).getIdFor() != null;
    }

    /**
     * 判断消息是否为请求
     * 对应TS: function isRequest(obj): boolean
     */
    public static boolean isRequest(Message message) {
        return message instanceof Request;
    }

    /**
     * 深度检查对象是否可JSON序列化
     * 对应TS: function isSerializableDeep(obj, seen = new WeakSet())
     */
    public static boolean isSerializableDeep(Object obj) {
        return isSerializableDeep(obj, new HashSet<>());
    }

    @SuppressWarnings("unchecked")
    private static boolean isSerializableDeep(Object obj, Set<Object> seen) {
        if (obj == null) {
            return true;
        }

        // 基本类型
        if (obj instanceof String || obj instanceof Number || obj instanceof Boolean) {
            return true;
        }

        // 增强类型: byte[], java.util.Date, java.math.BigInteger
        if (obj instanceof byte[] || obj instanceof java.util.Date || obj instanceof java.math.BigInteger) {
            return enhanceType;
        }

        // 检查List
        if (obj instanceof List) {
            if (seen.contains(obj)) {
                return false; // 循环引用
            }
            seen.add(obj);
            for (Object item : (List<Object>) obj) {
                if (!isSerializableDeep(item, seen)) {
                    return false;
                }
            }
            return true;
        }

        // 检查Map (对应JS plain object)
        if (obj instanceof Map) {
            if (seen.contains(obj)) {
                return false; // 循环引用
            }
            seen.add(obj);
            for (Map.Entry<?, ?> entry : ((Map<?, ?>) obj).entrySet()) {
                if (!(entry.getKey() instanceof String)) {
                    return false; // JSON键必须是字符串
                }
                if (!isSerializableDeep(entry.getValue(), seen)) {
                    return false;
                }
            }
            return true;
        }

        // 其他类型 (类实例等) - 不可序列化
        return false;
    }

    // ==================== 断言/校验工具 ====================

    /**
     * 校验Request
     * 对应TS: function assertRequest(request)
     */
    public static void assertRequest(Request request) {
        if (request == null) {
            throw new RuntimeException("request: expected object but got null");
        }
        if (request.getId() == null || !(request.getId() instanceof String)) {
            throw new RuntimeException("request.id: expected string, got " + request.getId());
        }
        if (request.getMeta() == null) {
            throw new RuntimeException("request.meta: expected object, got null");
        }
        if (request.getMethod() == null) {
            throw new RuntimeException("request.method: expected string, got null");
        }
        if (request.getObjectId() == null) {
            throw new RuntimeException("request.objectId: expected string, got null");
        }
        if (request.getArgs() == null) {
            throw new RuntimeException("request.args: expected array, got null");
        }
    }

    /**
     * 校验请求参数是否可序列化
     * 对应TS: function assertArgJSON(request)
     */
    public static void assertArgJSON(Request request) {
        for (int i = 0; i < request.getArgs().size(); i++) {
            Object arg = request.getArgs().get(i);
            if (!isSerializableDeep(arg)) {
                System.err.println("Non-serializable arg: " + arg);
                throw new RuntimeException(request.getMethod() + ".args[" + i + "] is not serializable");
            }
        }
    }

    /**
     * 校验Response
     * 对应TS: function assertResponse(response)
     */
    public static void assertResponse(Response response) {
        if (response == null) {
            throw new RuntimeException("response: expected object but got null");
        }
        if (response.getId() == null) {
            throw new RuntimeException("response.id: expected string, got null");
        }
        if (response.getIdFor() == null) {
            throw new RuntimeException("response.idFor: expected string, got null");
        }
    }

    /**
     * 从Map构建ProxyDescriber (用于JSON反序列化后的转换)
     */
    @SuppressWarnings("unchecked")
    public static ProxyDescriber mapToProxyDescriber(Map<String, Object> map) {
        return ProxyDescriber.fromMap(map);
    }
}

File: xuri-rpc-java/src/main/kotlin/com/xuri/rpc/Rpc.kt
@file:Suppress("UNCHECKED_CAST")

package com.xuri.rpc

import kotlinx.coroutines.*
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicLong

/**
 * RPC框架核心工具类 (Kotlin object单例)
 * 对应TS中 rpc.ts 模块级别的变量和函数
 *
 * 定时任务使用 CoroutineScope + delay 实现，天然支持协程
 */
object Rpc {

    // ==================== 全局状态 ====================

    /** 调试标志, 对应TS: let debugFlag = false */
    @Volatile
    var debugFlag: Boolean = false

    /** 本地主机ID, 对应TS: let hostId: string | null = null */
    @Volatile
    var hostId: String? = null
        private set

    /** 全局消息接收器, 对应TS: let messageReceiver */
    @Volatile
    private var messageReceiver: MessageReceiver? = null

    /** ID计数器, 对应TS: let idCount = 0 */
    private val idCount = AtomicLong(0)

    /** 增强类型标志, 对应TS: let enchanceType = true */
    var enhanceType: Boolean = true

    /** GC超时限制(毫秒), 对应TS: const gcTimeLimit = 30 * 1000 */
    const val GC_TIME_LIMIT = 30_000L

    /** 连接超时限制(毫秒), 对应TS: const connectionLimit = 30 * 1000 */
    const val CONNECTION_LIMIT = 30_000L

    /**
     * 每个hostId的选项映射
     * 对应TS: let options: Record<string|symbol, MessageReceiverOptions> = {}
     */
    private val options = ConcurrentHashMap<String, MessageReceiverOptions>()

    /** 默认host的key, 对应TS: let defaultHost = Symbol('defaultHost') */
    private const val DEFAULT_HOST_KEY = "__DEFAULT_HOST__"

    // ==================== 浅自动包装器 ====================

    /**
     * 浅层自动包装器
     * 对应TS: const shallowAutoWrapper: AutoWrapper
     *
     * 规则:
     * - null → 原样
     * - RpcFunction → 包装为代理 (对应TS typeof == 'function')
     * - List中含RpcFunction → 整个List包装为代理
     * - Map中含RpcFunction值 → 整个Map包装为代理
     * - 其他 → 原样
     */
    val shallowAutoWrapper: AutoWrapper = { obj ->
        when {
            obj == null -> null
            obj is RpcFunction -> asProxy(obj)
            obj is List<*> -> {
                if (obj.any { it is RpcFunction }) asProxy(obj) else obj
            }
            obj is Map<*, *> -> {
                if (obj.values.any { it is RpcFunction }) asProxy(obj) else obj
            }
            else -> obj
        }
    }

    // ==================== 公共方法 ====================

    /**
     * 设置主机ID
     * 对应TS: export function setHostId(id: string)
     */
    fun setHostId(id: String) {
        hostId = id
        getOrCreateOption(null).hostId = id
    }

    /**
     * 按ID删除代理
     * 对应TS: export function _deleteProxyById(id, hostId?)
     */
    fun deleteProxyById(id: String, hostIdParam: String? = null) {
        getOrCreateOption(hostIdParam).objectOfProxyManager.deleteById(id)
    }

    /**
     * 按对象删除代理
     * 对应TS: export function _deleteProxy(obj, hostId?)
     */
    fun deleteProxy(obj: Any, hostIdParam: String? = null) {
        getOrCreateOption(hostIdParam).objectOfProxyManager.delete(obj)
    }

    /**
     * 生成唯一ID
     * 对应TS: function getId() { return hostId + '' + (idCount++) }
     */
    fun getId(): String = "${hostId ?: ""}${idCount.getAndIncrement()}"

    /**
     * 获取或创建全局消息接收器
     * 对应TS: export function getMessageReceiver()
     */
    fun getMessageReceiver(): MessageReceiver {
        return messageReceiver ?: synchronized(this) {
            messageReceiver ?: MessageReceiver().also { messageReceiver = it }
        }
    }

    /**
     * 获取或创建指定hostId的选项
     * 对应TS: function getOrCreateOption(id?)
     */
    fun getOrCreateOption(id: String?): MessageReceiverOptions {
        val key = when {
            id == null -> DEFAULT_HOST_KEY
            id == hostId -> DEFAULT_HOST_KEY
            else -> id
        }
        return options.getOrPut(key) {
            MessageReceiverOptions(if (key == DEFAULT_HOST_KEY) null else key)
        }
    }

    // ==================== 代理相关 ====================

    /**
     * 获取或生成对象的代理ID
     * 对应TS: function getOrGenerateObjectId(obj, hostIdFrom)
     */
    private fun getOrGenerateObjectId(obj: Any, hostIdFrom: String?): String {
        val proxyManager = getOrCreateOption(hostIdFrom).objectOfProxyManager
        checkNotNull(hostId) { "hostId is null" }

        if (!proxyManager.has(obj)) {
            proxyManager.set(obj, getId())
        }
        return proxyManager.get(obj)!!
    }

    /**
     * 为对象创建代理描述
     * 对应TS: function createProxyForObject(proxyId, obj, hostId)
     */
    private fun createProxyForObject(proxyId: String, obj: Any?, hostIdParam: String?): ProxyDescriber? {
        if (obj == null) return null

        val members = mutableListOf<ProxyDescriber.MemberDescriber>()

        when (obj) {
            // 对应TS: typeof obj == 'function'
            is RpcFunction -> {
                members.add(ProxyDescriber.MemberDescriber("function", "__call__"))
            }
            // 对应TS: Object.keys(obj).filter(k => typeof obj[k] == 'function').filter(k => !k.startsWith('__'))
            is Map<*, *> -> {
                (obj as Map<String, Any?>).forEach { (key, value) ->
                    if (value is RpcFunction && !key.startsWith("__")) {
                        members.add(ProxyDescriber.MemberDescriber("function", key))
                    }
                }
            }
            // 非Map非RpcFunction的对象: 可通过反射获取方法 (扩展点)
        }

        return ProxyDescriber(id = proxyId, hostId = hostIdParam ?: "", members = members)
    }

    /**
     * 将对象包装为代理PreArgObj
     * 对应TS: export function asProxy(obj, hostIdFrom?)
     */
    fun asProxy(obj: Any?, hostIdFrom: String? = null): PreArgObj {
        val resolvedHostId = getOrCreateOption(hostIdFrom).hostId
        val id = getOrGenerateObjectId(obj!!, resolvedHostId)
        val proxy = createProxyForObject(id, obj, resolvedHostId)
        return PreArgObj("proxy", proxy)
    }

    /**
     * 生成错误响应
     * 对应TS: export function generateErrorReply(message, errorText, status = 500)
     */
    fun generateErrorReply(message: Request, errorText: String, status: Int = 500): Response =
        Response(id = getId(), idFor = message.id, trace = errorText, status = status)

    // ==================== 定时任务 (协程版) ====================

    /**
     * 启动自动检查定时任务
     * 对应TS: export function autoCheck()
     *
     * 使用 CoroutineScope + delay 替代 setInterval:
     * - killTimeoutConnection: 清理超时连接
     * - autoReRegister: 自动重新注册 (suspend)
     * - removeOutdatedProxyObject: 移除过期代理
     */
    fun autoCheck(scope: CoroutineScope = CoroutineScope(Dispatchers.Default + SupervisorJob())) {
        scope.launch {
            while (isActive) {
                delay(3000)
                killTimeoutConnection()
            }
        }
        scope.launch {
            while (isActive) {
                delay(3000)
                try {
                    autoReRegister() // suspend: 内部可能调用远程方法
                } catch (e: Exception) {
                    System.err.println("autoReRegister failed: ${e.message}")
                }
            }
        }
        scope.launch {
            while (isActive) {
                delay(3000)
                removeOutdatedProxyObject()
            }
        }
    }

    /**
     * 清理超时连接
     * 对应TS: function killTimeoutConnection(client?, millSec?)
     */
    fun killTimeoutConnection(client: Client? = null, millSec: Long = -1) {
        val timeout = if (millSec == -1L) CONNECTION_LIMIT else millSec

        if (client != null) {
            processTimeoutForOption(getOrCreateOption(client.getHostId()), timeout)
        } else {
            options.values.forEach { processTimeoutForOption(it, timeout) }
        }
    }

    private fun processTimeoutForOption(option: MessageReceiverOptions, timeout: Long) {
        val reqPending = option.requestPendingDict
        val toRemove = mutableListOf<String>()
        for ((key, pending) in reqPending) {
            if (System.currentTimeMillis() - pending.sendTime > timeout) {
                pending.deferred.completeExceptionally(RuntimeException("timeout"))
                toRemove.add(key)
            }
        }
        toRemove.forEach { reqPending.remove(it) }
    }

    /**
     * 自动重新注册远程代理
     * 对应TS: export function autoReRegister()
     *
     * 此方法是 suspend 的，因为内部调用 client.getObject() (suspend)
     */
    suspend fun autoReRegister() { // 对应TS中的隐式async (调用了async getObject)
        for (option in options.values) {
            val manager = option.runnableProxyManager

            for ((client, ids) in manager.clientMap) {
                if (ids.isNullOrEmpty()) continue

                val toReRegister = mutableListOf<List<Any?>>()
                val toRemove = mutableSetOf<String>()

                for (id in ids) {
                    val obj = manager.get(id)
                    if (obj == null) {
                        toRemove.add(id)
                    } else {
                        toReRegister.add(listOf(id))
                    }
                }

                // 清理已失效的id
                ids.removeAll(toRemove)
                if (ids.isEmpty()) {
                    manager.removeFromClientMap(client)
                }

                // 调用远程 reRegister
                if (toReRegister.isNotEmpty()) {
                    try {
                        val main0 = client.getObject("main0") // suspend
                        if (main0 is RemoteObject) {
                            main0.invoke("reRegister", toReRegister) // suspend
                        }
                    } catch (e: Exception) {
                        System.err.println("autoReRegister call failed: ${e.message}")
                    }
                }
            }
        }
    }

    /**
     * 移除过期代理对象
     * 对应TS: export function removeOutdatedProxyObject(timeout = -1)
     * (已修复bug: 正确遍历 reverseProxyMap)
     */
    fun removeOutdatedProxyObject(timeout: Long = -1) {
        val effectiveTimeout = if (timeout <= 0) GC_TIME_LIMIT else timeout
        for ((key, option) in options) {
            val manager = option.objectOfProxyManager
            val before = manager.size
            val count = manager.removeOutdated(effectiveTimeout)
            if (count > 0 && debugFlag) {
                println("$key removed $count proxies, before $before after ${manager.size}")
            }
        }
    }

    // ==================== 类型检查工具 ====================

    /**
     * 判断是否为简单/原始类型
     * 对应TS: function isSimpleObject(obj)
     */
    fun isSimpleObject(obj: Any?): Boolean =
        obj == null || obj is String || obj is Number || obj is Boolean || obj is ByteArray

    /**
     * 判断消息是否为响应
     * 对应TS: function isResponse(message)
     */
    fun isResponse(message: Message): Boolean =
        message is Response && message.idFor != null

    /**
     * 深度检查对象是否可JSON序列化
     * 对应TS: function isSerializableDeep(obj, seen?)
     */
    fun isSerializableDeep(obj: Any?, seen: MutableSet<Any> = mutableSetOf()): Boolean {
        if (obj == null) return true
        if (obj is String || obj is Number || obj is Boolean) return true

        // 增强类型
        if (obj is ByteArray || obj is java.util.Date || obj is java.math.BigInteger) {
            return enhanceType
        }

        // List
        if (obj is List<*>) {
            if (obj in seen) return false
            seen.add(obj)
            return obj.all { isSerializableDeep(it, seen) }
        }

        // Map (对应JS plain object)
        if (obj is Map<*, *>) {
            if (obj in seen) return false
            seen.add(obj)
            return obj.all { (k, v) -> k is String && isSerializableDeep(v, seen) }
        }

        return false
    }

    // ==================== 断言/校验 ====================

    fun assertRequest(request: Request) {
        requireNotNull(request.id) { "request.id: expected string" }
        requireNotNull(request.meta) { "request.meta: expected object" }
        requireNotNull(request.method) { "request.method: expected string" }
        requireNotNull(request.objectId) { "request.objectId: expected string" }
        requireNotNull(request.args) { "request.args: expected array" }
    }

    fun assertArgJSON(request: Request) {
        request.args.forEachIndexed { i, arg ->
            if (!isSerializableDeep(arg)) {
                System.err.println("Non-serializable arg: $arg")
                throw RuntimeException("${request.method}.args[$i] is not serializable")
            }
        }
    }

    fun assertResponse(response: Response) {
        requireNotNull(response.id) { "response.id: expected string" }
        requireNotNull(response.idFor) { "response.idFor: expected string" }
    }
}
